storage/remote/client.go

```diff
@@ -148,30 +151,41 @@ func (c *Client) Read(ctx context.Context, query *prompb.Query) (*prompb.QueryRe
 	if err != nil {
 		return nil, fmt.Errorf("error sending request: %v", err)
 	}
-	defer httpResp.Body.Close()
+
 	if httpResp.StatusCode/100 != 2 {
+		io.Copy(ioutil.Discard, httpResp.Body)
+		httpResp.Body.Close()
 		return nil, fmt.Errorf("server returned HTTP status %s", httpResp.Status)
 	}
 
-	compressed, err = ioutil.ReadAll(httpResp.Body)
-	if err != nil {
-		return nil, fmt.Errorf("error reading response: %v", err)
-	}
-
-	uncompressed, err := snappy.Decode(nil, compressed)
-	if err != nil {
-		return nil, fmt.Errorf("error reading response: %v", err)
-	}
+	querySet := DecodeReadResponse(httpResp.Header, httpResp.Body)
+
+	// TODO(bplotka): Do not buffer response in concreteSeriesSet. Add caller support for remote.QuerySet or
```

Comment:
It feels like we should attack this later. We need to ensure callers supports query set or serieSet that you should always `empty` in case of any non series error.